core/sail/inferencer/src/main/java/org/eclipse/rdf4j/sail/inferencer/fc/ForwardChainingSchemaCachingRDFSInferencer.java

```diff
@@ -0,0 +1,318 @@
+import org.eclipse.rdf4j.sail.NotifyingSail;
+import org.eclipse.rdf4j.sail.SailConflictException;
+import org.eclipse.rdf4j.sail.SailException;
+import org.eclipse.rdf4j.sail.inferencer.InferencerConnection;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.StampedLock;
+import java.util.stream.Collectors;
+
+/**
+ * @author Håvard Mikkelsen Ottestad
+ */
+
+public class ForwardChainingSchemaCachingRDFSInferencer extends AbstractForwardChainingInferencer {
+
+    // The schema, or null
+    Repository schema;
+
+    private static final Random random = new Random();
+
+    private StampedLock readWriteLock = new StampedLock();
+
+    // If false, the inferencer will skip some RDFS rules.
+    boolean useAllRdfsRules = true;
+
+    // Tbox cache
+    Collection<Resource> properties = new HashSet<>();
+    Collection<Resource> types = new HashSet<>();
+    Collection<Statement> subClassOfStatements = new ArrayList<>();
+    Collection<Statement> subPropertyOfStatements = new ArrayList<>();
+    Collection<Statement> rangeStatements = new ArrayList<>();
+    Collection<Statement> domainStatements = new ArrayList<>();
+
+    // Forward chained Tbox cache as lookup tables
+    Map<Resource, Set<Resource>> calculatedTypes = new HashMap<>();
+    Map<Resource, Set<Resource>> calculatedProperties = new HashMap<>();
+    Map<Resource, Set<Resource>> calculatedRange = new HashMap<>();
+    Map<Resource, Set<Resource>> calculatedDomain = new HashMap<>();
+
+    // The inferencer has been instantiated from another inferencer and shares it's schema with that one
+    private boolean sharedSchema;
+
+    void clearInferenceTables() {
+        properties = new HashSet<>();
+        types = new HashSet<>();
+        subClassOfStatements = new ArrayList<>();
+        subPropertyOfStatements = new ArrayList<>();
+        rangeStatements = new ArrayList<>();
+        domainStatements = new ArrayList<>();
+
+        calculatedTypes = new HashMap<>();
+        calculatedProperties = new HashMap<>();
+        calculatedRange = new HashMap<>();
+        calculatedDomain = new HashMap<>();
+    }
+
+    void readLock(ForwardChainingSchemaCachingRDFSInferencerConnection connection) {
+
+//        if (numberOfThreadsWaitingForWriteLock.get() > 0) {
+//            System.err.println("starve reads");
+//        }
+
+        while (numberOfThreadsWaitingForWriteLock.get() > 0) {
+            Thread.yield();
+        }
+
+        if (connection.lockStamp != 0) {
+            throw new SailException("Connection already has a lock! Might be a begin() without a commit() or rollback() for previous transaction().");
+        }
+        connection.lockStamp = readWriteLock.readLock();
```

Comment:
Not good practice to be manipulating the connections instance variables from here. Better done using a specific class. Have you looked into the LockManager utilities that we have already created in the past? 

Interface is the following, and there are a few different implementations of it:

https://github.com/eclipse/rdf4j/blob/a283e758fb79e7b62e94b669547e6e64c64106cf/core/sail/api/src/main/java/org/eclipse/rdf4j/sail/LockManager.java